File: 948-Bag-of-Token/ts/main.ts
function bagOfTokensScore(tokens: number[], power: number): number {
    tokens.sort((a, b) => a - b);

    let left = 0;
    let right = tokens.length - 1;

    let score = 0;
    let max_score = 0;

    while (left <= right) {
        if (tokens[left] <= power) {
            power -= tokens[left];
            left++;
            score++;
            max_score = Math.max(max_score, score);
        } else if (score > 0) {
            power += tokens[right];
            right--;
            score--;
        } else {
            break;
        }
    }

    return max_score;
}

File: 948-Bag-of-Token/go/main.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func main() {
	fmt.Println("Hello World")
	tokens := []int{100, 200, 300, 400}
	power := 200
	fmt.Println(bagOfTokensScore(tokens, power))
}

func bagOfTokensScore(tokens []int, power int) int {
	sort.Ints(tokens)

	left := 0
	right := len(tokens) - 1

	score := 0
	max_score := 0

	for left <= right {
		if tokens[left] <= power {
			power -= tokens[left]
			left += 1
			score += 1
			max_score = int(math.Max(float64(max_score), float64(score)))
		} else if score > 0 {
			power += tokens[right]
			right -= 1
			score -= 1
		} else {
			break
		}
	}

	return max_score
}
